ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/worldview/BonsaiArchiver.java

```diff
@@ -0,0 +1,243 @@
+                                (slotKey, slotValue) -> {
+                                  // Move any previous state for this account
+                                  archivedAccountStorageCount.addAndGet(
+                                      rootWorldStateStorage.archivePreviousStorageState(
+                                          blockchain.getBlockHeader(
+                                              blockchain
+                                                  .getBlockHeader(blockHash)
+                                                  .get()
+                                                  .getParentHash()),
+                                          Bytes.concatenate(
+                                              address.addressHash(), slotKey.getSlotHash())));
+                                });
+                          });
+                }
+                LOG.atDebug()
+                    .setMessage("All account state and storage archived for block {}")
+                    .addArgument(block.getKey())
+                    .log();
+                rootWorldStateStorage.setLatestArchivedBlock(block.getKey());
+
+                // Update local var for logging progress
+                latestArchivedBlock.set(block.getKey());
+                if (latestArchivedBlock.get() % 100 == 0) {
+                  // Log progress in case catching up causes there to be a large number of keys
+                  // to move
+                  LOG.atInfo()
+                      .setMessage(
+                          "archive progress: state up to block {} archived ({} behind chain head {})")
+                      .addArgument(latestArchivedBlock.get())
+                      .addArgument(blockchain.getChainHeadBlockNumber() - latestArchivedBlock.get())
+                      .addArgument(blockchain.getChainHeadBlockNumber())
+                      .log();
+                }
+              });
+
+      LOG.atDebug()
+          .setMessage(
+              "finished moving state for blocks {} to {}. Archived {} account state entries, {} account storage entries")
+          .addArgument(blocksToArchive.firstKey())
+          .addArgument(latestArchivedBlock.get())
+          .addArgument(archivedAccountStateCount.get())
+          .addArgument(archivedAccountStorageCount.get())
+          .log();
+    }
+
+    return archivedAccountStateCount.get() + archivedAccountStorageCount.get();
+  }
+
+  private final Lock archiveMutex = new ReentrantLock(true);
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent addedBlockContext) {
+    initialize();
+    final Optional<Long> blockNumber =
+        Optional.of(addedBlockContext.getBlock().getHeader().getNumber());
+    blockNumber.ifPresent(
+        blockNum -> {
+          // Since moving blocks can be done in batches we only want
+          // one instance running at a time
+          executeAsync.accept(
+              () -> {
+                if (archiveMutex.tryLock()) {
```

Comment:
With `synchronized` in `moveBlockStateToArchive`, do you need to add this lock here. I guess the idea is to be sure that we finished moving the block N state to archive before block N+1, `moveBlockStateToArchive` should handle that case.